doc/auto_storage.md

```diff
@@ -564,29 +565,134 @@ or re-formatted.
 The storage proposal must make possible to define what to do with existing partitions and logical
 volumes. Even with existing MD RAIDs or LVM volume groups.
 
-In order to provide the same capabilities than the Guided proposal (see below) it must be possible
-to specify that a given partition (or LVM logical volume) must be:
+A `search` section allows to match the definition of a partition or an LVM logical volume with one
+(or several) devices existing in the system. In order to provide the same capabilities than the
+Guided proposal (see below) it must be possible to specify that a given partition or volume must be:
 
   - Deleted if needed to make space for the newly defined devices
   - Deleted in all cases
   - Shrunk to the necessary size to make space for new devices
-  - Shrunk to a given size, maybe a range, in all cases (not really possible in the current Guided
+  - Shrunk or extended to a given size, maybe a range (not really possible in the current Guided
     Proposal)
 
-It would be desirable to have the possibility of expressing some combinations of the above, like
-"try to shrink it to make space but proceed to delete it if shrinking it is not enough".
+It is even possible to express some combinations of the above, like "try to shrink it to make space
+but proceed to delete it if shrinking it is not enough".
+
+Deletion can be achieved with the corresponding `delete` flag or the alternative `deleteIfNeeded`.
+If any of those flags are active for a partition, it makes no sense to specify any other usage
+(like declaring a file system on it).
+
+The following example deletes the partition with the label "root" in all cases and, if needed, keeps
+deleting other partitions as needed to make space for the new partition of 30 GiB.
+
+```json
+"storage": {
+    "drives": [
+        {
+            "partitions": [
+                {
+                    "search": {
+                        "condition": { "property": "fsLabel", "value": "root" }
+                    },
+                    "delete": true
+                },
+                { "search": {}, "deleteIfNeeded": true },
+                { "size": "30 GiB" }
+            ]
+        }
+    ]
+}
+```
+
+Often some partitions or logical volumes are shrunk only to make space for the declared devices. But
+since resizing is not a destructive operation, it can also make sense to declare a given partition
+must be resized (shrunk or extended) and then formatted and/or mounted.
+
+In any case, note that resizing a partition can be limited depending on its content, the filesystem
+type, etc.
+
+Combining `search` and `resize` is enough to indicate Agama is expected to resize a given partition
```

Comment:
Let's elaborate a bit in the most technical part how all this could be achieved.

Let's start with a small reminder on how resizing works in the current SpaceMaker.

```ruby
class Proposal::PartitionsDistributionCalculator
  # [...]
  # Space that should be freed when resizing an existing partition in order to have a good chance
  # of creating a valid PartitionsDistribution (by means of #best_distribution).
  #
  # It does not take the limitations of the partition into account (eg. whether it can be
  # resized or how full it is)
  def resizing_size
    # This tries to create valid distributions
    # If resizing does not result in any valid distribution, returns current size
  end
end

class Y2Storage::Proposal::SpaceMakerActions::Shrink
  # @return [DiskSize] size of the space to substract ideally
  attr_accessor :shrink_size

  # Reduces the size of the target partition
  #
  # If possible, it reduces the size of the partition by {#shrink_size}.
  # Otherwise, it reduces the size as much as possible.
  # 
  # This method does not take alignment into account.
  def shrink(devicegraph)
     [...]
  end
end
```

Taking that into account, the following steps/changes could be inserted in the general `AgamaProposal` algorithm to handle all the reasonable situations.

First of all, Beware the usage of "current" can lead to strange situations like:
  - Real size: 10 GiB. Min: "current". Max: 5 GiB
Those cases should be watched. The steps below assume that does not happen.

**Step 1.** When calculating the SpaceMaker actions, generate one shrink for every partition in which min is smaller than the current size.

How to determine the size of the shrink action?
```ruby
suggested = current - calculator.resizing_size
target_size =
if min > suggested
    min
elsif max < suggested
    max
else
    suggested
end
```

The action is mandatory if the max is smaller than the current size. Otherwise is optional.

**Step 2.** To be done on every attempt. That is:

 - Before starting with the optional actions and also after every action
 - Before calculating the corresponding distribution
 - For all partitions that need to grow (that can be optimized, not all partitions are affected by each action)

```ruby
def needs_to_grow?(partition)
  partition.min > device_at_working_devicegraph(partition.found_device).size
end

if needs_to_grow?(partition)
  immediately_try_to_grow_up_to_min(partition)
  # Maybe it would be useful to remember the result
  #
  # It can fail because there is still not enough space, then that distribution is not valid
  # and we need to keep trying stuff.
  #
  # It can fail because there was space but something else at resize_info prevents
  # the partition from growing.
  # Should we abort? I guess so, that's what "min" means after all.
end
```

**Step 3.** When distributing extra space at the end of the algorithm

Take into account partitions that could_grow and are right before the space

```ruby
def could_grow?(partition)
  partition.max > partition.current && partition.resize_info.max_size > partition.current
end
```

A general notes regarding the order in which we decide what to delete: maybe we should prioritize to delete partitions that are after a partition that needs to grow. If I remember correctly, the only criteria is starting by the end of the disk, but it could be way smarter.